src/components/SearchableSingleSelect/SearchableSingleSelect.tsx

```diff
@@ -0,0 +1,175 @@
+}: {
+    msg: string
+    onRetryClick: () => void
+}) {
+    return (
+        <div className={classes.error}>
+            <div className={classes.errorInnerWrapper}>
+                <span className={classes.loadingErrorLabel}>{msg}</span>
+                <button
+                    className={classes.errorRetryButton}
+                    type="button"
+                    onClick={onRetryClick}
+                >
+                    {i18n.t('Retry')}
+                </button>
+            </div>
+        </div>
+    )
+}
+
+type OnChange = ({ selected }: { selected: string }) => void
+type OnFilterChange = ({ value }: { value: string }) => void
+type OnIntersectionChange = ({
+    isIntersecting,
+}: {
+    isIntersecting: boolean
+}) => void
+interface SearchableSingleSelectPropTypes {
+    onChange: OnChange
+    onFilterChange: OnFilterChange
+    onIntersectionChange: OnIntersectionChange
+    onRetryClick: () => void
+    options: Option[]
+    preventIntersectionDetection: boolean
+    showEndLoader: boolean
+    loading: boolean
+    selected?: string
+    error?: string
+}
+
+export const SearchableSingleSelect = ({
+    error,
+    loading,
+    onChange,
+    onFilterChange,
+    onIntersectionChange,
+    options,
+    preventIntersectionDetection,
+    selected,
+    showEndLoader,
+    onRetryClick,
+}: SearchableSingleSelectPropTypes) => {
+    const [currentlyIntersecting, setCurrentlyIntersecting] = useState(false)
+    const [loadingSpinnerRef, setLoadingSpinnerRef] = useState<HTMLElement>()
+    const debouncedOnFilterChange = useDebouncedCallback<OnFilterChange>(
+        (args) => onFilterChange(args),
+        200
+    )
+
+    // We want to defer the actual filter value so we don't send a request with
+    // every key stroke
+    const [filterValue, _setFilterValue] = useState('')
+    const setFilterValue = useCallback(
+        (nextFilterValue: string) => {
+            _setFilterValue(nextFilterValue)
+            debouncedOnFilterChange({ value: nextFilterValue })
+        },
+        [debouncedOnFilterChange]
+    )
+
+    useEffect(() => {
+        // We don't want to wait for intersections when loading as that can
+        // cause buggy behavior
+        if (loadingSpinnerRef && !preventIntersectionDetection) {
+            const observer = new IntersectionObserver(
+                (entries) => {
+                    const [{ isIntersecting }] = entries
+
+                    if (isIntersecting !== currentlyIntersecting) {
+                        setCurrentlyIntersecting(isIntersecting)
+                        onIntersectionChange({ isIntersecting })
+                    }
+                },
+                { threshold: 0.8 }
+            )
+
+            observer.observe(loadingSpinnerRef)
+            return () => observer.disconnect()
+        }
+    }, [
+        loadingSpinnerRef,
+        currentlyIntersecting,
+        preventIntersectionDetection,
+        onIntersectionChange,
+    ])
```

Comment:
I don't think `currentlyIntersecting` is needed. We only trigger this observer once, since it's recreated due to `Loading` unmounting. Also, if we're just interested in when " end is reached", I think it makes sense to not call the callback when `isIntersecting` is false. 

So I think it could be simplified like this:

```ts
    useEffect(() => {
        // We don't want to wait for intersections when loading as that can
        // cause buggy behavior
        if (!loadingSpinnerRef || preventIntersectionDetection) {
            return
        }
        const observer = new IntersectionObserver(
            (entries) => {
                const [{ isIntersecting }] = entries
                if (isIntersecting) {
                    onIntersectionChange({ isIntersecting })
                    observer.disconnect() //might not need this, at least from my testing this is not called twice with `isIntersecting: true` 
                }
            },
            { threshold: 0.8 }
        )

        observer.observe(loadingSpinnerRef)
        return () => observer.disconnect()
    }, [loadingSpinnerRef, preventIntersectionDetection, onIntersectionChange])
```

Also prefer the early return, because it wasn't obvious to me at first that the cleanup was inside the "if". That is subjective though.

If you don't link this, I think we also could make `currentlyIntersecting` a ref instead.